modules/standard/Version.chpl

```diff
@@ -201,61 +159,37 @@ module Version {
       return 1;
   }
 
-  private proc spaceship(v1: semanticVersion(?),
-                         v2: semanticVersion(?)) param : int {
+  private proc spaceship(v1: sourceVersion(?),
+                         v2: sourceVersion(?)) param : int {
     param majComp = spaceship(v1.major, v2.major);
-    if majComp != 0 then
+    if majComp != 0 {
       return majComp;
-    param minComp = spaceship(v1.minor, v2.minor);
-    if minComp != 0 then
-      return minComp;
-    return spaceship(v1.update, v2.update);
-  }
-
-
-  // Comparisons between semanticVersions
-
-  proc ==(v1: semanticVersion(?), v2: semanticVersion(?)) param : bool {
-    return spaceship(v1,v2) == 0;
-  }
-
-  /*
-    Equality/inequality operators between two values of type
-    :type:`semanticVersion` check whether or not the two values
-    have identical major, minor, and update values.
-  */
-  proc !=(v1: semanticVersion(?), v2: semanticVersion(?)) param : bool {
-    return spaceship(v1,v2) != 0;
-  }
-
-  proc <(v1: semanticVersion(?), v2: semanticVersion(?)) param : bool {
-    return spaceship(v1,v2) < 0;
-  }
-
-  proc <=(v1: semanticVersion(?), v2: semanticVersion(?)) param : bool {
-    return spaceship(v1,v2) <= 0;
-  }
-
-  proc >(v1: semanticVersion(?), v2: semanticVersion(?)) param : bool {
-    return spaceship(v1,v2) > 0;
-  }
-
-  /*
-    Ordered comparisons between two values of type
-    :type:`semanticVersion` are based on the lexical ordering of each
-    value's ``major.minor.update`` value, reflecting an "was released
-    before/after" relationship.
-  */
-  proc >=(v1: semanticVersion(?), v2: semanticVersion(?)) param : bool {
-    return spaceship(v1,v2) >= 0;
+    } else {
+      param minComp = spaceship(v1.minor, v2.minor);
+      if minComp != 0 {
+        return minComp;
+      } else {
+        param upComp = spaceship(v1.update, v2.update);
+        if upComp != 0 {
+          return upComp;
+        } else if v1.commit == v2.commit {
+          return 0;
+        } else if v1.commit == "" {
+          return 1;
+        } else if v2.commit == "" then {
+          return -1;
+        } else {
+          // sentinel for "not comparable"
+          return 2;
```

Comment:
I found this confusing. I wonder if an enum would be cleaner. Don't need to see it for this PR given the release deadlines, but looking at the callsites I found this confusing.

Reply:
I agree it's not as clear as it could be, and really wanted to return a tuple of (answer, ambiguous) where answer would be -1, 0, or 1, and ambiguous would be true or false.  But then we get into tuples of params, which seemed... scary to rely upon (maybe unfairly).

A param enum could also work, though it's less spaceship-y and seems like it results in a similar amount of special-casing (or maybe you see a plan for making it clean that I'm not?).

Or really, having it throw and the callsites catch would be clean, but definitely not supported at compile-time.

Reply:
(I could also move completely away from using the spaceship helper, but think your previous suggestion to use it was generally a good one that helped remove a lot of redundant code from the individual operators).